crates/core/tedge_api/src/messages.rs

```diff
@@ -616,6 +616,47 @@ impl ConfigUpdateCmdPayload {
     }
 }
 
+#[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub struct FirmwareMetadata {
+    pub name: Option<String>,
+    pub version: Option<String>,
+    #[serde(rename = "url")]
+    pub remote_url: Option<String>,
```

Comment:
Generally if the url of the firmware binary is known then it should be included. If you are talking about publishing the current firmware info (name, version) on startup, then it is possible that the url is not known (as it might be baked into the OS image and was not provided via a url).

Reply:
Since the `url` is useless in the initial bootstrap message, that reports the factory firmware info on the device, the only other case where it'd be useful is after a firmware update is performed. In that case, this `url` must be the one the device received in the cloud request (SR 515)? I don't see why that's needed either, because the `successful` status message already includes that `remoteUrl` field, which the mapper is using to send the `115` message. So, including the same `url` here seems redundant.

Reply:
I don't have the context anymore of what message we're talking about.

Reply:
But having the `url` property can't harm the message, and it is more descriptive about where the firmware came from (if the info is available).

Reply:
> I don't have the context anymore of what message we're talking about. But more information can be useful..but without concrete examples of what messages we're explicitly talking about then I can't make an informed decision.

Okay, so here are the two cases where this "tedge firmware metadata" message would be sent:

1. When a device is getting bootstrapped: it needs to publish its factory firmware details. In this case, at best the device would know its `version` and `name`. So, the device is less likely to use the `url` field in this path.
2. After the device firmware is updated using a cloud request (with a remote URL): the `firmware_update` request would contain that `remoteUrl` and once this operation is successful, the device would send the `successful` command status message with the same `remoteUrl`. The mapper on receipt of this `successful` status will mark the forward that to the cloud (SR 503 message), followed by the SmartREST set firmware message (SR 115) where the same `remoteUrl` in the `successful` status response is used. Since the firmware version in the cloud is already updated this way, the device need not send the **tedge firmware metadata message** again. Now, on the next restart of the agent, this tedge firmware metadata message would be sent again, but here also, it won't be able to send that `url` information, unless it persists that URL info after every firmware update, which is less likely/unnecessary IMO.

This is the reason why I feel that the `url` field in the "tedge firmware metadata message" is completely useless and it is relevant only in the **firmware_update command** messages.

> But having the url property can't harm the message, and it is more descriptive about where the firmware came from (if the info is available).

Yeah, having an extra field doesn't harm, but when we document this API, we'll have to give some guidance to the user on when/how to use it, right? If this field is never going to be used, as neither cases highlighted above had a need for it (without additional persistence requirements), why even have it as an optional field?

Reply:
I think we should still keep it as it "could" be useful to people if they want to implement it (e.g. workflow overriding etc.).

Reply:
I'd have taken the opposite approach of omitting it for now and adding it when we encounter the need for it, as it's always easy to add things to a public API than removing things. Even for the workflow overriding, I only see the `remoteUrl` field in the firmware update `cmd` message being used and not this `url` field in the capability message.

But, since you're fairly confident that it would be useful in some use-cases, and since it's an optional field anyway, let's keep it for now. Once firmware update is fully implemented, and if this field is still not used, we can still remove it before 1.0.